### ISSUE TITLE
training custom datastes not converge

### ISSUE THREAD

**Issue by idontlikelongname** [Megvii-BaseDetection/YOLOX#98]
I followed the tutorials to train my custom datasets with 7 classes. After 300 epoch that accoding to default settings. I find that the output of fg_conf is very low. 
Demo.py results is empty with conf_thresh equal to 0.5, even 0.1 is empty.
attached by training logs and exp files.

exs. the conf_loss is high, you can see it from the attached log file.

```
from yolox.exp import Exp as MyExp
from yolox.data import get_yolox_datadir


class Exp(MyExp):
    def __init__(self):
        super(Exp, self).__init__()
        # self.num_classes = 7
        self.num_classes = 10
        self.depth = 1.33
        self.width = 1.25
        self.exp_name = os.path.split(os.path.realpath(__file__))[1].split(".")[0]
        self.max_epoch = 300
        self.eval_interval = self.max_epoch

    def get_data_loader(self, batch_size, is_distributed, no_aug=False):
        from yolox.data import (
            BaiduDataset,
            TrainTransform,
            YoloBatchSampler,
            DataLoader,
            InfiniteSampler,
            MosaicDetection,
        )

        dataset = BaiduDataset(
            data_dir=os.path.join('/root/datasets/paddle/train'),
            img_size=self.input_size,
            preproc=TrainTransform(
                rgb_means=(0.485, 0.456, 0.406),
                std=(0.229, 0.224, 0.225),
                max_labels=50,
            ),
        )

        dataset = MosaicDetection(
            dataset,
            mosaic=not no_aug,
            img_size=self.input_size,
            preproc=TrainTransform(
                rgb_means=(0.485, 0.456, 0.406),
                std=(0.229, 0.224, 0.225),
                max_labels=120,
            ),
            degrees=self.degrees,
            translate=self.translate,
            scale=self.scale,
            shear=self.shear,
            perspective=self.perspective,
            enable_mixup=self.enable_mixup,
        )

        self.dataset = dataset

        if is_distributed:
            batch_size = batch_size // dist.get_world_size()

        sampler = InfiniteSampler(
            len(self.dataset), seed=self.seed if self.seed else 0
        )

        batch_sampler = YoloBatchSampler(
            sampler=sampler,
            batch_size=batch_size,
            drop_last=False,
            input_dimension=self.input_size,
            mosaic=not no_aug,
        )

        dataloader_kwargs = {"num_workers": self.data_num_workers, "pin_memory": True}
        dataloader_kwargs["batch_sampler"] = batch_sampler
        train_loader = DataLoader(self.dataset, **dataloader_kwargs)

        return train_loader

    def get_eval_loader(self, batch_size, is_distributed, testdev=False):
        from yolox.data import BaiduDataset, ValTransform

        valdataset = BaiduDataset(
            data_dir=os.path.join('/root/datasets/paddle/train'),
            img_size=self.test_size,
            preproc=ValTransform(
                rgb_means=(0.485, 0.456, 0.406),
                std=(0.229, 0.224, 0.225),
            ),
        )

        if is_distributed:
            batch_size = batch_size // dist.get_world_size()
            sampler = torch.utils.data.distributed.DistributedSampler(
                valdataset, shuffle=False
            )
        else:
            sampler = torch.utils.data.SequentialSampler(valdataset)

        dataloader_kwargs = {
            "num_workers": self.data_num_workers,
            "pin_memory": True,
            "sampler": sampler,
        }
        dataloader_kwargs["batch_size"] = batch_size
        val_loader = torch.utils.data.DataLoader(valdataset, **dataloader_kwargs)

        return val_loader

    def get_evaluator(self, batch_size, is_distributed, testdev=False):
        from yolox.evaluators import VOCEvaluator

        val_loader = self.get_eval_loader(batch_size, is_distributed, testdev=testdev)
        evaluator = VOCEvaluator(
            dataloader=val_loader,
            img_size=self.test_size,
            confthre=self.test_conf,
            nmsthre=self.nmsthre,
            num_classes=self.num_classes,
        )
        return evaluator
[train_log.txt](https://github.com/Megvii-BaseDetection/YOLOX/files/6866489/train_log.txt)

```

**Comment by idontlikelongname** [comment]
And more, this is my datasets class

```
class BaiduDataset(Dataset):
    """
    Baidu dataset class.
    """

    def __init__(
        self,
        data_dir=None,
        name="baidu_v1",
        img_size=(640, 640),
        preproc=None,
    ):
        """
        Baidu dataset initialization.
        Args:
            data_dir (str): dataset root directory
            name (str): COCO data name (e.g. 'train2017' or 'val2017')
            img_size (int): target image size after pre-processing
            preproc: data augmentation strategy
        """
        super().__init__(img_size)
        self.name = name
        self.img_size = img_size
        self.preproc = preproc
        self.DET_TYPES = [1, 2, 3, 4, 5, 6, 7]
        # self.DET_TYPES = [3]
        self.SEG_TYPES = [8, 9, 10]

        assert data_dir is not None
        self.data_dir = data_dir

        sequence_file = sorted(os.listdir(os.path.join(self.data_dir, 'image')))
        sequence_file = [name.replace('.jpg', '') for name in sequence_file]
        self.ids = sequence_file

        self._load_anno_all()

    def _load_anno_all(self):
        annotations_all = []
        for id_ in self.ids:
            # load anno
            anno_json_file = os.path.join(self.data_dir, 'annotation', '{}.json'.format(id_))
            with open(anno_json_file, 'r') as f:
                anno_infos = json.load(f)

            # load labels
            annotations = anno_infos['labels']
            height, width = anno_infos['img_shape']
            valid_objs = []
            for obj in annotations:
                if obj['type'] in self.DET_TYPES:
                    x1 = np.max((0, obj["x"]))
                    y1 = np.max((0, obj["y"]))
                    x2 = np.min((width - 1, x1 + np.max((0, obj['width'] - 1))))
                    y2 = np.min((height - 1, y1 + np.max((0, obj['height'] - 1))))
                    if x2 > x1 and y2 > y1:
                        obj["clean_bbox"] = [x1, y1, x2, y2]
                        valid_objs.append(obj)
            if len(valid_objs) == 0:
                continue

            objs = valid_objs
            num_objs = len(objs)

            res = np.zeros((num_objs, 5))

            for ix, obj in enumerate(objs):
                cls = self.DET_TYPES.index(obj["type"])
                res[ix, 0:4] = obj["clean_bbox"]
                res[ix, 4] = cls
            img_info = (height, width)
            annotations_all.append(dict(res=res, img_info=img_info, id=id_, img_path=os.path.join(
                self.data_dir, 'image', '{}.jpg'.format(id_))))
        self.annotations_all = annotations_all
        assert len(self.annotations_all) <= len(self.ids)

    def __len__(self):
        return len(self.annotations_all)

    def stats_info(self):
        name_counts = dict()
        for i in range(len(self.annotations_all)):
            annotations = self.annotations_all[i]
            for oi in range(len(annotations['res'])):
                box_type = annotations['res'][oi, -1]
                if BAIDU_CLASSES[int(box_type-1)] not in name_counts:
                    name_counts[BAIDU_CLASSES[int(box_type-1)]] = 1
                else:
                    name_counts[BAIDU_CLASSES[int(box_type-1)]] += 1
        for name, count in name_counts.items():
            print("++", name, '\t', count)

    def load_anno(self, index):
        assert index < len(self.annotations_all)
        annotations = self.annotations_all[index]

        return annotations['res']

    def pull_item(self, index):
        annotations = self.annotations_all[index]

        id_ = annotations['id']
        img_file = annotations['img_path']

        img = cv2.imread(img_file)
        assert img is not None
        height, width, _ = img.shape

        # load anno
        img_info = (height, width)

        annotations = self.annotations_all[index]
        assert img_info == annotations['img_info']

        return img, annotations['res'], annotations['img_info'], annotations['id']

    @Dataset.resize_getitem
    def __getitem__(self, index):
        """
        One image / label pair for the given index is picked up and pre-processed.

        Args:
            index (int): data index

        Returns:
            img (numpy.ndarray): pre-processed image
            padded_labels (torch.Tensor): pre-processed label data.
                The shape is :math:`[max_labels, 5]`.
                each label consists of [class, xc, yc, w, h]:
                    class (float): class index.
                    xc, yc (float) : center of bbox whose values range from 0 to 1.
                    w, h (float) : size of bbox whose values range from 0 to 1.
            info_img : tuple of h, w, nh, nw, dx, dy.
                h, w (int): original shape of the image
                nh, nw (int): shape of the resized image without padding
                dx, dy (int): pad size
            img_id (int): same as the input index. Used for evaluation.
        """
        img, res, img_info, img_id = self.pull_item(index)

        if self.preproc is not None:
            img, target = self.preproc(img, res, self.input_dim)
        else:
            target = None
        return img, target, img_info, img_id



```

**Comment by ruinmessi** [comment]
You can decrease your lr. 
Your training loss is actually very low at the beginning of training but raised up at the final. Maybe you can also reduce the total training epoch from 300 to 100 since the size of your dataset is quite small.
